mllib/src/main/scala/org/apache/spark/ml/feature/RobustScaler.scala

```diff
@@ -147,49 +146,40 @@ class RobustScaler (override val uid: String)
 
   override def fit(dataset: Dataset[_]): RobustScalerModel = {
     transformSchema(dataset.schema, logging = true)
-    val localRelativeError = $(relativeError)
 
-    val summaries = dataset.select($(inputCol)).rdd.map {
-      case Row(vec: Vector) => vec
-    }.mapPartitions { iter =>
-      var agg: Array[QuantileSummaries] = null
-      while (iter.hasNext) {
-        val vec = iter.next()
-        if (agg == null) {
-          agg = Array.fill(vec.size)(
-            new QuantileSummaries(QuantileSummaries.defaultCompressThreshold, localRelativeError))
-        }
-        require(vec.size == agg.length,
-          s"Number of dimensions must be ${agg.length} but got ${vec.size}")
-        var i = 0
-        while (i < vec.size) {
-          agg(i) = agg(i).insert(vec(i))
-          i += 1
-        }
-      }
-
-      if (agg == null) {
-        Iterator.empty
-      } else {
-        Iterator.single(agg.map(_.compress))
-      }
-    }.treeReduce { (agg1, agg2) =>
-      require(agg1.length == agg2.length)
-      var i = 0
-      while (i < agg1.length) {
-        agg1(i) = agg1(i).merge(agg2(i))
-        i += 1
-      }
-      agg1
+    val numFeatures = MetadataUtils.getNumFeatures(dataset, $(inputCol))
+    val vectors = dataset.select($(inputCol)).rdd.map {
+      case Row(vec: Vector) =>
+        require(vec.size == numFeatures,
+          s"Number of dimensions must be $numFeatures but got ${vec.size}")
+        vec
     }
 
-    val (range, median) = summaries.map { s =>
-      (s.query($(upper)).get - s.query($(lower)).get,
-        s.query(0.5).get)
-    }.unzip
-
-    copyValues(new RobustScalerModel(uid, Vectors.dense(range).compressed,
-      Vectors.dense(median).compressed).setParent(this))
+    val localRelativeError = $(relativeError)
+    val localUpper = $(upper)
+    val localLower = $(lower)
+
+    val (ranges, medians) = vectors.mapPartitions { iter =>
+      if (iter.hasNext) {
+        iter.flatMap { vec => Iterator.range(0, numFeatures).map(i => (i, vec(i))) } ++
+          // add NaN for each feature as the indication of the end of partition
+          Iterator.range(0, numFeatures).map((_, Double.NaN))
+      } else Iterator.empty
+    }.aggregateByKey(
+      new QuantileSummaries(QuantileSummaries.defaultCompressThreshold, localRelativeError))(
+      // compress the QuantileSummaries at the end of partition
+      seqOp = (s, v) => if (v.isNaN) s.compress else s.insert(v),
```

Comment:
I don't think you'll necessarily get the NaNs at the end, right?  you'll get multiple NaNs too, so if you did get them all at the end you would compress repeatedly. (I wonder if compress can be smarter about returning itself if there's no work to do, not sure)

You might happen to get the values for keys by partition though in practice, which would work out well if true. Is that what this is kind of hoping for?

Reply:
I append multiple NaNs at the end of each partions, since there are numFeatures aggregators on each partition. One NaN is for a aggregator to trigger a compress.

Reply:
After trigger compression at the _map_ side, I do not need to confirm compression at merge `combOp` and the final `query` in the previous commits.

Reply:
I think my mental model of aggregateByKey might be wrong or something, but what are you guaranteed about the order in which seqOp encounters the values from flatMap previously? after a shuffle collects all the values for a feature and starts aggregating a partition of some values, you might get the NaNs first or in the middle or something. Worst case they're all at the start and you dont' end up compressing the final result at all. It would make sense to me if it happened to shuffle the values in order, and you'd encounter (..., NaN, ...values..., NaN, ...values..., NaN) as it would aggregate the values as they came out of flatMap in order. I'm not sure if there's a better way to affect what seqOp does "last" though, so maybe it's the best that can be done.

Reply:
There is no guarantee in the api doc of `aggregateByKey` or somewhere about the order in which seqOp encounters the values. I just dig into its impl and think it should work like this. So I agree that it is not a nice way to trigger compression like this.
I guess adding a new RDD operation like `aggregateByKeyWithinPartitions` maybe a better choice.

I will remove the usage of NaN and confirm comression before merge and query.